failed to parse protocol - http

<!-- Thanks for reporting a bug for this project. READ THIS FIRST:

- Provide as many details as possible. Simply saying "X gives bug" or "X gives error" is not enough!
- Paste logs, configuration sample and code into the backticks (```).
- Read through the template carefully and fill out all missing details.
- Please also search for similar issues in this issue tracker first and read through the ESPHome FAQ.

DO NOT DELETE ANY TEXT from this template! Otherwise the issue may be closed without a comment.
-->

**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

PIP esphome-dev  

**ESP (ESP32/ESP8266, Board/Sonoff):**

ESP32

**Affected component:**

```http_request```

**Description of problem:**

Error about the http url parsing.

**Problem-relevant YAML-configuration entries:**
```yaml
binary_sensor:
  - platform: esp32_touch
    name: "ESP32 Touch Pad GPIO27"
    pin: GPIO27
    threshold: 1000
    on_press:
      then:
        - http_request.post:
            url: http://online.oliveiradigital.net/user.php?cont=login&username=u123&password=u123&Submit=Login&md5=123
            verify_ssl: false  

```

**Logs (if applicable):**
```
[12:11:41][D][binary_sensor:036]: 'ESP32 Touch Pad GPIO27': Sending state ON
[12:11:41][E][HTTPClient.cpp:241] beginInternal(): failed to parse protocol
[12:11:41][E][HTTPClient.cpp:241] beginInternal(): failed to parse protocol
[12:11:41][W][http_request:032]: HTTP Request failed at the begin phase. Please check the configuration
```

**Additional information and things you've tried:**
Seems to come from here: https://github.com/espressif/arduino-esp32/blob/master/libraries/HTTPClient/src/HTTPClient.cpp#L139

I get this warning while compiling:

```
src/esphome/components/http_request/http_request.cpp: In member function 'const char* esphome::http_request::HttpRequestComponent::get_string()':
src/esphome/components/http_request/http_request.cpp:81:89: warning: function may return address of local variable [-Wreturn-local-addr]
 const char *HttpRequestComponent::get_string() { return this->client_.getString().c_str(); }
                                                                                         ^
src/esphome/components/http_request/http_request.cpp:81:81: note: declared here
 const char *HttpRequestComponent::get_string() { return this->client_.getString().c_str(); }
```


<!-- LEAVE THIS LINE AS-IS AND DON'T DELETE IT, OTHERWISE THE ISSUE WILL BE CLOSED AUTOMATICALLY. -->

--- a/esphome/components/http_request/http_request.cpp
+++ b/esphome/components/http_request/http_request.cpp
@@ -15,15 +15,15 @@ void HttpRequestComponent::dump_config() {
 void HttpRequestComponent::send() {
   bool begin_status = false;
   this->client_.setReuse(true);
-  static const String url = this->url_.c_str();
+  static const String URL = this->url_.c_str();
 #ifdef ARDUINO_ARCH_ESP32
-  begin_status = this->client_.begin(url);
+  begin_status = this->client_.begin(URL);
 #endif
 #ifdef ARDUINO_ARCH_ESP8266
 #ifndef CLANG_TIDY
   this->client_.setFollowRedirects(true);
   this->client_.setRedirectLimit(3);
-  begin_status = this->client_.begin(*this->get_wifi_client_(), url);
+  begin_status = this->client_.begin(*this->get_wifi_client_(), URL);
 #endif
 #endif
 
@@ -44,7 +44,8 @@ void HttpRequestComponent::send() {
 
   int http_code = this->client_.sendRequest(this->method_, this->body_.c_str());
   if (http_code < 0) {
-    ESP_LOGW(TAG, "HTTP Request failed; URL: %s; Error: %s", this->url_.c_str(), HTTPClient::errorToString(http_code).c_str());
+    ESP_LOGW(TAG, "HTTP Request failed; URL: %s; Error: %s", this->url_.c_str(),
+             HTTPClient::errorToString(http_code).c_str());
     this->status_set_warning();
     return;
   }
@@ -80,8 +81,8 @@ WiFiClient *HttpRequestComponent::get_wifi_client_() {
 void HttpRequestComponent::close() { this->client_.end(); }
 
 const char *HttpRequestComponent::get_string() {
-  static const String str = this->client_.getString();
-  return str.c_str();
+  static const String STR = this->client_.getString();
+  return STR.c_str();
 }
 
 }  // namespace http_request

--- a/esphome/components/http_request/http_request.cpp
+++ b/esphome/components/http_request/http_request.cpp
@@ -15,14 +15,15 @@ void HttpRequestComponent::dump_config() {
 void HttpRequestComponent::send() {
   bool begin_status = false;
   this->client_.setReuse(true);
+  static const String url = this->url_.c_str();
 #ifdef ARDUINO_ARCH_ESP32
-  begin_status = this->client_.begin(this->url_);
+  begin_status = this->client_.begin(url);
 #endif
 #ifdef ARDUINO_ARCH_ESP8266
 #ifndef CLANG_TIDY
   this->client_.setFollowRedirects(true);
   this->client_.setRedirectLimit(3);
-  begin_status = this->client_.begin(*this->get_wifi_client_(), this->url_);
+  begin_status = this->client_.begin(*this->get_wifi_client_(), url);
 #endif
 #endif
 
@@ -43,19 +44,19 @@ void HttpRequestComponent::send() {
 
   int http_code = this->client_.sendRequest(this->method_, this->body_.c_str());
   if (http_code < 0) {
-    ESP_LOGW(TAG, "HTTP Request failed; URL: %s; Error: %s", this->url_, HTTPClient::errorToString(http_code).c_str());
+    ESP_LOGW(TAG, "HTTP Request failed; URL: %s; Error: %s", this->url_.c_str(), HTTPClient::errorToString(http_code).c_str());
     this->status_set_warning();
     return;
   }
 
   if (http_code < 200 || http_code >= 300) {
-    ESP_LOGW(TAG, "HTTP Request failed; URL: %s; Code: %d", this->url_, http_code);
+    ESP_LOGW(TAG, "HTTP Request failed; URL: %s; Code: %d", this->url_.c_str(), http_code);
     this->status_set_warning();
     return;
   }
 
   this->status_clear_warning();
-  ESP_LOGD(TAG, "HTTP Request completed; URL: %s; Code: %d", this->url_, http_code);
+  ESP_LOGD(TAG, "HTTP Request completed; URL: %s; Code: %d", this->url_.c_str(), http_code);
 }
 
 #ifdef ARDUINO_ARCH_ESP8266
@@ -78,7 +79,10 @@ WiFiClient *HttpRequestComponent::get_wifi_client_() {
 
 void HttpRequestComponent::close() { this->client_.end(); }
 
-const char *HttpRequestComponent::get_string() { return this->client_.getString().c_str(); }
+const char *HttpRequestComponent::get_string() {
+  static const String str = this->client_.getString();
+  return str.c_str();
+}
 
 }  // namespace http_request
 }  // namespace esphome

--- a/esphome/components/http_request/http_request.h
+++ b/esphome/components/http_request/http_request.h
@@ -28,7 +28,7 @@ class HttpRequestComponent : public Component {
   float get_setup_priority() const override { return setup_priority::AFTER_WIFI; }
 
   void set_url(std::string url) {
-    this->url_ = url.c_str();
+    this->url_ = url;
     this->secure_ = url.compare(0, 6, "https:") == 0;
   }
   void set_method(const char *method) { this->method_ = method; }
@@ -42,7 +42,7 @@ class HttpRequestComponent : public Component {
 
  protected:
   HTTPClient client_{};
-  const char *url_;
+  std::string url_;
   const char *method_;
   const char *useragent_{nullptr};
   bool secure_;

